Prototope/ShapeLayer.swift

```diff
@@ -0,0 +1,424 @@
+	}
+	
+	
+	/** Creates a rectangle with an optional corner radius. */
+	convenience public init(rectangle: Rect, cornerRadius: Double = 0, parent: Layer? = nil, name: String? = nil) {
+		var renderRect = rectangle
+		renderRect.origin = Point()
+		self.init(segments: Segment.segmentsForRect(renderRect, cornerRadius: cornerRadius), closed: true, parent: parent, name: name)
+		self.frame = rectangle
+	}
+	
+	
+	/** Creates a line from two points. */
+	convenience public init(lineFromFirstPoint firstPoint: Point, toSecondPoint secondPoint: Point, parent: Layer? = nil, name: String? = nil) {
+		let difference = secondPoint - firstPoint
+		self.init(segments: Segment.segmentsForLineFromFirstPoint(Point(), secondPoint: difference), parent: parent, name: name)
+		self.frame = Rect(x: firstPoint.x, y: firstPoint.y, width: difference.x, height: difference.y)
+	}
+	
+	
+	/** Creates a regular polygon path with the given number of sides. */
+	convenience public init(polygonCenteretAtPoint centerPoint: Point, radius: Double, numberOfSides: Int, parent: Layer? = nil, name: String? = nil) {
+		let frame = Rect(x: centerPoint.x - radius, y: centerPoint.y - radius, width: radius * 2, height: radius * 2)
+		self.init(segments: Segment.segmentsForPolygonCenteredAtPoint(frame.center, radius: radius, numberOfSides: numberOfSides), closed: true, parent: parent, name: name)
+		self.frame = frame
+	}
+	
+	
+	/** Initialize the ShapeLayer with a given path. */
+	public init(segments: [Segment], closed: Bool = false, parent: Layer? = nil, name: String? = nil) {
+		
+		self.segments = segments
+		self.closed = closed
+		
+		super.init(parent: parent, name: name, viewClass: ShapeView.self)
+		
+		self.shapeViewLayer.path = self.bezierPath.CGPath
+		self.shapeViewLayerStyleChanged()
+	}
+	
+	
+	// MARK: - Segments
+	
+	/** A list of all segments of this path. */
+	public var segments: [Segment]
+	
+	/** Gets the first segment of the path, if it exists. */
+	public var firstSegment: Segment? {
+		return segments.first
+	}
+	
+	
+	/** Gets the last segment of the path, if it exists. */
+	public var lastSegment: Segment? {
+		return segments.last
+	}
+	
+	
+	/** Convenience method to add a point by wrapping it in a segment. */
+	public func addPoint(point: Point) {
+		self.segments.append(Segment(point: point))
+	}
+	
+	
+	/** Replaces the segment at the given index. Throws an environment exception if the given index isn't in the segment list. */
+	public func replaceSegmentAtIndex(index: Int, withSegment segment: Segment) {
+		if index >= self.segments.count {
+			Environment.currentEnvironment?.exceptionHandler("Tried to replace a path segment at index \(index) but there are only \(self.segments.count) elements")
+			return
+		}
+		
+		self.segments[index] = segment
+		
+	}
+	
+	
+	// TODO(jb): How can this be triggered automatically when mutating the segments?
+	/** Redraws the path. You can call this after you change path segments. */
+	public func setNeedsDisplay() {
```

Comment:
This is what I'm not sure of. I made it public because if they mutate the list of segments on their own, then they need to inform we need a redraw (following paper's design here).

Could instead make this read only and only let you mutate through Approved Methods™ instead?